tidb-lightning: add glossary and other updates (#1693)
pingcap/docs: dev/reference/tools/tidb-lightning/glossary.md

---
title: TiDB Lightning Glossary

> summary: List of special terms used in TiDB Lightning.

Reviewer: How about adding these TiDB Lightning terms to the existing `glossary.md` at the root directory? (`docs/dev/glossary.md`) I think maintaining one glossary would be friendlier to users. They can easily search all terms in one place.

Author: 🤔 pingcap/docs-cn#2066 (DM) placed it separately already, and https://github.com/pingcap/docs-cn/pull/2102#pullrequestreview-327970130 (Binlog) is pending change. WDYT @pingcap/ecosystem-tools.

Reviewer 2: How about adding Lightning & Binlog terms to the existing `glossary.md`, but keeping DM separately.

Reviewer: @csuzhangxc Yes. DM will have its own separate TOC later on the website, so we can maintain a separate glossary for DM.

Reviewer 3: I don't recommend putting glossary  together for two reasons
* proper nouns are not uniformly planned
* proper nouns are more suitable in their own fields，but can be refered, such as backup and recovery toolsets, rather than uniform placement. Obviously many proper nouns only make sense in its domain

Reviewer: - Words in the glossary will be written and added based on our current documentation.
- I agree with you that proper nouns only make sense in a specific field. That's why it's a proper noun and why we need to explain it. If users cannot get what it means in what context, then we need to improve that description to make it clearer.

"be unified" is never only for "unity". Instead, we want to provide one place for users to look up any word that they are not sure about its meaning.

Reviewer 3: Let me summarize
* you concern that how to look up any word
* I concern that how user get the context (the specific field information) 

So how do we resolve it, I don't think these two goals are contradictory

Author: @lilin90  @GregoryIan How are we going to resolve this? This PR and pingcap/docs-cn#2102 are still waiting for the decision...

Reviewer: @kennytm Changes in this PR are acceptable. We can add them to the complete Glossary by providing links later.

<!-- B -->

## B

### Back end

The back end is the destination where TiDB Lightning sends the parsed result to. Sometimes it is also spelled as "backend".

See [TiDB Lightning "TiDB" Back End](/dev/reference/tools/tidb-lightning/tidb-backend.md) for details.

### Black-white list

A configuration list which specifies which tables should be imported, and which should be excluded.

See [TiDB Lightning Table Filter](/dev/reference/tools/tidb-lightning/table-filter.md) for details.

<!-- C -->

## C

### Checkpoint

TiDB Lightning continuously saves its progress into a local file or remote database while importing. This allows TiDB Lightning to resume from middle if it spuriously crashed. See the [Checkpoints](/dev/reference/tools/tidb-lightning/checkpoints.md) section for details.

### Checksum

In TiDB Lightning, the checksum of a table is a set of 3 numbers calculated from the content of each KV pair in that table. These are respectively:

* the number of KV pairs,
* total length of all KV pairs, and
* the bitwise-XOR of [CRC-64-ECMA](https://en.wikipedia.org/wiki/Cyclic_redundancy_check) value each pair.

> TiDB Lightning [validates the imported data](/dev/faq/tidb-lightning.md#how-to-ensure-the-integrity-of-the-imported-data) by comparing the [local](/dev/reference/tools/tidb-lightning/glossary.md#local-checksum) and [remote checksums](/dev/reference/tools/tidb-lightning/glossary.md#remote-checksum) of every table. The program would stop if any pair is not equal. This check can be skipped by setting the configuration `post-restore.checksum` to `false`.

Reviewer 4: ```suggestion
TiDB Lightning [validates the imported data](/dev/faq/tidb-lightning.md#how-to-ensure-the-integrity-of-the-imported-data) by comparing the [local](/dev/reference/tools/tidb-lightning/glossary.md#local-checksum) and [remote checksums](/dev/reference/tools/tidb-lightning/glossary.md#remote-checksum) of every table. The program would stop if any pair does not match. You can skip this check by setting the `post-restore.checksum` configuration to `false`.
```

See also the [Troubleshooting guide](/dev/how-to/troubleshoot/tidb-lightning.md#checksum-failed-checksum-mismatched-remote-vs-local) for how to properly handle checksum mismatch.

### Chunk

Equivalent to a single file in the data source.

### Compaction

> An operation which merges multiple small SST files into a large SST file, and cleans up deleted entries.

Reviewer 4: ```suggestion
An operation that merges multiple small SST files into one large SST file, and cleans up the deleted entries.
```

> 

Reviewer 4: ```suggestion
```

> TiKV automatically compacts data in background while TiDB Lightning is importing. Due to legacy reasons, you can configure TiDB Lightning to explicitly trigger a compaction every time a table is imported. This is now discouraged and the corresponding settings are disabled by default.

Reviewer 4: ```suggestion
TiKV automatically compacts data in background while TiDB Lightning is importing.
> **Note:**
>
>  For legacy reasons, you can still configure TiDB Lightning to explicitly trigger a compaction every time a table is imported. However, this is not recommended and the corresponding settings are disabled by default.
```

See [RocksDB's wiki page on Compaction](https://github.com/facebook/rocksdb/wiki/Compaction) for its technical details.

<!-- D -->

## D

### Data engine

An [engine](/dev/reference/tools/tidb-lightning/glossary.md#engine) for sorting actual row data.

> When a table is very large, its data is placed into multiple data engines to improve task pipelining and save space of TiKV Importer. By default a new data engine is opened every 100 GB of SQL data, which can be configured through the `mydumper.batch-size` setting.

Reviewer 4: ```suggestion
When a table is very large, its data is placed into multiple data engines to improve task pipelining and save space of TiKV Importer. By default, a new data engine is opened for every 100 GB of SQL data, which can be configured through the `mydumper.batch-size` setting.
```

TiDB Lightning processes multiple data engines concurrently. This is controlled by the `lightning.table-concurrency` setting.

<!-- E -->

## E

### Engine

In TiKV Importer, an engine is a RocksDB instance for sorting KV pairs.

> TiDB Lightning transfers data to TiKV Importer through engines. It first opens an engine, sends lots of KV pairs to it (with no particular order), and finally closes the engine. The engine will sort the received KV pairs after it is closed. These closed engines can then be further uploaded to the TiKV stores for ingestion.

Reviewer 4: ```suggestion
TiDB Lightning transfers data to TiKV Importer through engines. It first opens an engine, sends KV pairs to it (with no particular order), and finally closes the engine. The engine sorts the received KV pairs after it is closed. These closed engines can then be further uploaded to the TiKV stores for ingestion.
```

> Engines use TiKV Importer's `import-dir` as temporary storage, which are sometimes referred as "engine files".

Reviewer 4: ```suggestion
Engines use TiKV Importer's `import-dir` as temporary storage, which are sometimes referred to as "engine files".
```

See also [data engine](/dev/reference/tools/tidb-lightning/glossary.md#data-engine) and [index engine](/dev/reference/tools/tidb-lightning/glossary.md#index-engine).

<!-- I -->

## I

### Import mode

> A configuration which optimizes TiKV for writing but reduces read speed and space utility.

Reviewer 4: ```suggestion
A configuration that optimizes TiKV for writing at the cost of degraded read speed and space usage.
```

> TiDB Lightning automatically switches into and out of import mode while running. However, if TiKV is somehow stuck in import mode, you can use `tidb-lightning-ctl` to [force them revert](/dev/faq/tidb-lightning.md#why-my-tidb-cluster-is-using-lots-of-cpu-resources-and-running-very-slowly-after-using-tidb-lightning) to [normal mode](/dev/reference/tools/tidb-lightning/glossary.md#normal-mode).

Reviewer 4: ```suggestion
TiDB Lightning automatically switches to and off the import mode while running. However, if TiKV gets stuck in import mode, you can use `tidb-lightning-ctl` to [force revert](/dev/faq/tidb-lightning.md#why-my-tidb-cluster-is-using-lots-of-cpu-resources-and-running-very-slowly-after-using-tidb-lightning) to [normal mode](/dev/reference/tools/tidb-lightning/glossary.md#normal-mode).
```

### Index engine

An [engine](/dev/reference/tools/tidb-lightning/glossary.md#engine) for sorting indices.

Regardless of number of indices, every table is associated with exactly one index engine.

> TiDB Lightning processes multiple index engines concurrently. This is controlled by the `lightning.index-concurrency` setting. Since every table has exactly one index engine, this also configures the maximum number of tables to process in at the same time.

Reviewer 4: ```suggestion
TiDB Lightning processes multiple index engines concurrently. This is controlled by the `lightning.index-concurrency` setting. Since every table has exactly one index engine, this also configures the maximum number of tables to process at the same time.
```

### Ingestion

An operation which inserts the entire content of an [SST file](/dev/reference/tools/tidb-lightning/glossary.md#sst-file) into the RocksDB (TiKV) store.

> Ingestion is a very fast operation compared with inserting each KV pair one-by-one. This operation is the basis of speed of TiDB Lightning.

Reviewer 4: ```suggestion
Ingestion is a very fast operation compared with inserting KV pairs one by one. This operation is the determinant factor for the performance of TiDB Lightning.
```

<!-- K -->

## K

### KV pair

Abbreviation of "key-value pair".

### KV encoder

A routine which parses SQL or CSV rows to KV pairs. Multiple KV encoders run in parallel to speed up processing.

<!-- L -->

## L

### Local checksum

The [checksum](/dev/reference/tools/tidb-lightning/glossary.md#checksum) of a table calculated by TiDB Lightning itself before sending the KV pairs to TiKV Importer.

<!-- N -->

## N

### Normal mode

The mode where [import mode](/dev/reference/tools/tidb-lightning/glossary.md#import-mode) is disabled.

<!-- P -->

## P

### Post-processing

The period of time after the entire data source is parsed and sent to TiKV Importer. TiDB Lightning is waiting for TiKV Importer to upload and [ingest](/dev/reference/tools/tidb-lightning/glossary.md#ingest) the [SST files](/dev/reference/tools/tidb-lightning/glossary.md#sst-file).

<!-- R -->

## R

### Remote checksum

The [checksum](/dev/reference/tools/tidb-lightning/glossary.md#checksum) of a table calculated by TiDB after it has been imported.

<!-- S -->

## S

### Scattering

> An operation which randomly reassigns the leader and members of a [region](/dev/glossary.md#regionpeerraft-group). Scattering ensures the imported data are distributed evenly among all TiKV stores, and thus reduces PD's work.

Reviewer 4: ```suggestion
An operation which randomly reassigns the leader and the peers of a [region](/dev/glossary.md#regionpeerraft-group). Scattering ensures that the imported data are distributed evenly among TiKV stores. This reduces stress on PD.
```

### Ingestion

An operation which inserts the entire content of an [SST file](/dev/reference/tools/tidb-lightning/glossary.md#sst-file) into the RocksDB (TiKV) store.

Ingestion is a very fast operation compared with inserting KV pairs one by one. This operation is the determinant factor for the performance of TiDB Lightning.

See [RocksDB's wiki page on Creating and Ingesting SST files](https://github.com/facebook/rocksdb/wiki/Creating-and-Ingesting-SST-files) for its technical details.

<!-- K -->

## K

### KV pair

Abbreviation of "key-value pair".

### KV encoder

A routine which parses SQL or CSV rows to KV pairs. Multiple KV encoders run in parallel to speed up processing.

<!-- L -->

## L

### Local checksum

The [checksum](/dev/reference/tools/tidb-lightning/glossary.md#checksum) of a table calculated by TiDB Lightning itself before sending the KV pairs to TiKV Importer.

<!-- N -->

## N

### Normal mode

The mode where [import mode](/dev/reference/tools/tidb-lightning/glossary.md#import-mode) is disabled.

<!-- P -->

## P

### Post-processing

> The period of time after the entire data source is parsed and sent to TiKV Importer. TiDB Lightning is waiting for TiKV Importer to upload and [ingest](/dev/reference/tools/tidb-lightning/glossary.md#ingest) the [SST files](/dev/reference/tools/tidb-lightning/glossary.md#sst-file).

Reviewer 5: ```suggestion
The period of time after the entire data source is parsed and sent to TiKV Importer. TiDB Lightning is waiting for TiKV Importer to upload and [ingest](/dev/reference/tools/tidb-lightning/glossary.md#ingestion) the [SST files](/dev/reference/tools/tidb-lightning/glossary.md#sst-file).
```

Or use "ingest" instead? The noun form is rarely used in the doc.

<!-- R -->

## R

### Remote checksum

The [checksum](/dev/reference/tools/tidb-lightning/glossary.md#checksum) of a table calculated by TiDB after it has been imported.

<!-- S -->

## S

### Scattering

> An operation which randomly reassigns the leader and the peers of a [region](/dev/glossary.md#regionpeerraft-group). Scattering ensures that the imported data are distributed evenly among TiKV stores. This reduces stress on PD.

Reviewer 4: ```suggestion
An operation that randomly reassigns the leader and the peers of a [Region](/dev/glossary.md#regionpeerraft-group). Scattering ensures that the imported data are distributed evenly among TiKV stores. This reduces stress on PD.
```

Reviewer 4: Region needs to be initial-capped(大写首字母)
